cmd/azure.go

```diff
@@ -0,0 +1,363 @@
+	}
+	return a.GetObjectInfo(bucket, object)
+}
+
+// CopyObject - Use Azure equivalent CopyBlob.
+func (a AzureObjects) CopyObject(srcBucket, srcObject, destBucket, destObject string, metadata map[string]string) (objInfo ObjectInfo, err error) {
+	err = a.client.CopyBlob(destBucket, destObject, a.client.GetBlobURL(srcBucket, srcObject))
+	if err != nil {
+		return objInfo, azureToObjectError(traceError(err), srcBucket, srcObject)
+	}
+	return a.GetObjectInfo(destBucket, destObject)
+}
+
+// DeleteObject - Use azure equivalent DeleteBlob.
+func (a AzureObjects) DeleteObject(bucket, object string) error {
+	err := a.client.DeleteBlob(bucket, object, nil)
+	if err != nil {
+		return azureToObjectError(traceError(err), bucket, object)
+	}
+
+	return nil
+}
+
+// ListMultipartUploads - Incomplete implementation, for now just return the prefix if it is an incomplete upload.
+func (a AzureObjects) ListMultipartUploads(bucket, prefix, keyMarker, uploadIDMarker, delimiter string, maxUploads int) (result ListMultipartsInfo, err error) {
+	// FIXME: Full ListMultipartUploads is not supported yet. It is supported just enough to help our client libs to
+	// support re-uploads.
+	result.MaxUploads = maxUploads
+	result.Prefix = prefix
+	result.Delimiter = delimiter
+	resp, err := a.ListObjectParts(bucket, prefix, prefix, 1, 1000)
+	if err != nil {
+		// In case ListObjectParts returns error, it would mean that no such incomplete upload exists on
+		// azure storage - in which case we return nil. This plays a role for mc and our client libs. i.e client
+		// does ListMultipartUploads to figure the previous uploadID to continue uploading from where it left off.
+		// If we return error the upload never starts and always returns error.
+		return result, nil
+	}
+	if len(resp.Parts) > 0 {
+		result.Uploads = []uploadMetadata{{prefix, prefix, time.Now().UTC(), ""}}
+	}
+	return result, nil
+}
+
+// NewMultipartUpload - Use Azure equivalent CreateBlockBlob.
+func (a AzureObjects) NewMultipartUpload(bucket, object string, metadata map[string]string) (uploadID string, err error) {
+	// Azure doesn't return a unique upload ID and we use object name in place of it. Azure allows multiple uploads to
+	// co-exist as long as the user keeps the blocks uploaded (in block blobs) unique amongst concurrent upload attempts.
+	// Each concurrent client, keeps its own blockID list which it can commit.
+	uploadID = object
+	return uploadID, azureToObjectError(traceError(a.client.CreateBlockBlob(bucket, object)))
+}
+
+// CopyObjectPart - Not implemented.
+func (a AzureObjects) CopyObjectPart(srcBucket, srcObject, destBucket, destObject string, uploadID string, partID int, startOffset int64, length int64) (info PartInfo, err error) {
+	return info, traceError(NotImplemented{})
+}
+
+// Encode partID+md5Hex to a blockID.
+func azureGetBlockID(partID int, md5Hex string) string {
+	return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%.5d.%s", partID, md5Hex)))
+}
```

Comment:
This is incorrect - ```fmt.Sprintf("%.5d.%s", partID, md5Hex))```
the format-specifier for partID should be "%06d", since there can be upto 100,000 uncommitted blocks and it the blockID as a whole is required to be equal length, so we need to pad the part number with leading zeros.